### PR TITLE
Drivers Script

### PR DESCRIPTION
**Pull request by alexdkellogg**
Create a driver script that allows: debiasing of the treatment effect; auto-selection of the optimal method. Fixed a typo (input to impute).

### CODE REVIEW

File: r_code/estimator_drivers.R [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+                          horizon = c(0, 4), effect_type = "mean",
+                          id_var = "entry",
+                          time_var = "period",
+                          treat_indicator = "treatperiod_0",
+                          outcome_var = "target",
+                          counterfac_var = "counter_factual") {
+  # Ensure the method names are entered appropriately.
+  stopifnot(method_names %in%
+    c("CausalImpact", "Gsynth", "SDID", "SDID_Uncon", "SCM"))
+  # Make sure horizon is positive. Some methods have 0 training error is why.
```
Comment (jeffreyoldham) [inline r_code/estimator_drivers.R:362]: Is the second sentence proper English?

File: r_code/estimator_drivers.R [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+                                outcome_var = "target",
+                                counterfac_var = "counter_factual") {
+  # Identify the treated entries.
+  treated_ids <- .IdentifyTreated(raw_data,
+    id_var = id_var,
+    treat_indicator = treat_indicator
+  )
+  num_treated <- treated_ids %>%
+    dplyr::distinct(!!as.name(id_var), .keep_all = T) %>%
+    dplyr::filter(Treated == 1) %>%
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:126]: The helper before, .IdentifyTreated, actually just creates a column "Treated" equal to 0 or 1. So they are identified but not extracted yet.

File: r_code/estimator_drivers.R [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+  # Ensure the method names are appropriately specified.
+  stopifnot(method_name %in%
+    c("CausalImpact", "Gsynth", "SDID", "SDID_Uncon", "SCM"))
+  # Use rules of thumb to try to create an independent placebo set.
+  indep_data <- .CreateIndepPlacebo(raw_data,
+    id_var = id_var,
+    time_var = time_var,
+    treat_indicator = treat_indicator,
+    outcome_var = outcome_var,
+    counterfac_var = counterfac_var
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:50]: I'll go back and update this (for other files, I presume) in a later PR.

File: r_code/estimator_drivers.R [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    treat_indicator = treat_indicator
+  )
+  num_treated <- treated_ids %>%
+    dplyr::distinct(!!as.name(id_var), .keep_all = T) %>%
+    dplyr::filter(Treated == 1) %>%
+    dplyr::count() %>%
+    dplyr::pull()
+  num_donors <- treated_ids %>%
+    dplyr::distinct(!!as.name(id_var), .keep_all = T) %>%
+    dplyr::filter(Treated == 0) %>%
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:131]: See above. Maybe the name IdentifyTreated is confusing, but it literally identifies the treated with a column.

File: r_code/estimator_drivers.R [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+    dplyr::count() %>%
+    dplyr::pull()
+  # Split the data cleanly if we have 5x donors relative to treated.
+  if (num_donors / num_treated >= 5) {
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:135]: I can just use an if else statement with some type of flag for now?

File: r_code/estimator_drivers.R [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+#'    placebo treated set.
+#'
+#' @return A subset of the full data tibbles sampled randomly, with all relevant
+#'    information for each selected observation over time.
+.SelectDonorIDs <- function(raw_data, treated_ids, id_var, num_slice,
+                            frac_treated) {
+  set_aside_ids <- treated_ids %>%
+    dplyr::filter(Treated == 0) %>%
+    dplyr::distinct(!!as.name(id_var)) %>%
+    dplyr::slice_sample(n = num_slice) %>%
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:196]: Good call. Added question clarifying (without).

File: r_code/estimator_drivers.R (hunk 1 above)
Comment (alexdkellogg) [reply r_code/estimator_drivers.R:362]: 'Tis not.

Comment (alexdkellogg) [reply r_code/estimator_drivers.R:362]: Though technically the third "sentence" haha.

**Comment by alexdkellogg** [comment]
Auto estimator and debiasing features merge.

**Pull request merged**